File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/billing/shared/billing-transaction-item.model.ts
import { BillItemPriceVM } from './billing-view-models';
import { Patient } from '../../patients/shared/patient.model';
import {
  NgForm,
  FormGroup,
  FormControl,
  Validators,
  FormBuilder,
  ReactiveFormsModule,
  ValidationErrors,
  ValidatorFn
} from '@angular/forms'
import { BillingTransaction } from './billing-transaction.model';
import { CommonValidators } from '../../shared/common-validator';

export class BillingTransactionItem {

  public BillingTransactionItemId: number = 0;
  public BillingTransactionId: number = null;
  public PatientId: number = 0;
  //Ashim: 4Jan2018
  //ProviderId is AssignedToDr. It is used only in BillingTransaction.
  //In LabItemRequisition. ProviderId is RequestedBy Dr. from billing/ doctor who gives lab order in Doctors
  public ProviderId: number = null;
  public ProviderName: string = null;
  public ServiceDepartmentId: number = 0;
  public ServiceDepartmentName: string = null;
  public ProcedureCode: string = null;
  public ItemId: number = 0;
  public ItemName: string = null;
  public Price: number = 0;
  public SAARCCitizenPrice: number = 0; //Yubraj: 16th May '19
  public ForeignerPrice: number = 0; //Yubraj: 16th May '19
  public InsForeignerPrice: number = 0;
  public Quantity: number = null;
  public SubTotal: number = null;
  public DiscountPercent: number = 0;
  public DiscountPercentAgg: number = 0;
  public DiscountAmount: number = 0;
  public Tax: number = 0;
  public TotalAmount: number = 0;
  public BillStatus: string = null;
  public RequisitionId: number = null;
  public RequisitionDate: string = null;
  public PaidDate: string = null;
  public Remarks: string = null;
  public ReturnStatus: boolean = null;
  public ReturnQuantity: number = null;
  public CounterId: number = 0;
  public CounterDay: string = null;
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public CancelRemarks: string = null;
  public ItemList: Array<BillItemPriceVM> = new Array<BillItemPriceVM>();  // array to map the items from different department
  public TaxPercent: number = 0;
  public CancelledBy: number = null;//add cancelled-on also if needed. which is now done from server side.
  public BillingPackageId: number = null;

  public IsPaid: boolean = true;
  //Below variable only for check duplicate item or not
  public IsDuplicateItem: boolean = false;

  public Patient: Patient = new Patient();
  public ServiceDepartment = null;///service deparatment comes as a part of .Include() in billing controller.

  public BillingTransactionItemValidator: FormGroup = null;
  public RequestedBy: number = null;
  public RequestedByName: string = "";
  public PatientVisitId: number = null;
  //only used to update BillStatus in BIL_BillItemRequisition table during PendingBills.
  public BillItemRequisitionId: number = null;
  //only used in client side.
  public IsTaxApplicable: boolean = true;
  public TaxableAmount: number = 0;
  public PatientName: string = null;//mahesh:10Mar'19-needed for fraction.

  public NonTaxableAmount: number = 0;//added: sud: 29May'18
  public PaymentReceivedBy: number = null;//added: sud: 29May'18
  public PaidCounterId: number = null;//added: sud: 29May'18

  public BillingType: string = null;//sud: 19Jun'18--eg: IP, OP.
  public RequestingDeptId: number = null;//used mostly for Inpatient segregation.

  public IsSelected: boolean = false;//sud:12May
  public IsValidSelDepartment: boolean = true;
  public IsValidSelItemName: boolean = true;
  public IsValidSelAssignedToDr: boolean = true;
  public IsValidSelRequestedByDr: boolean = true;

  public BillItemPriceId: number = 0;//mahesh:10Mar'19-needed for fraction.

  public RequestingUserNameNDept: string = null;//sud:24Sept'18 -- needed for IP-Lab Request

  public VisitType: string = null;//sud: 27Sept'18 -- needed for Labs.

  //not mapped field - required while updating adt item quantity
  public ItemIntegrationName: string = null;
  //not mapped field - required while updating adt item quantity
  public SrvDeptIntegrationName: string = null;

  public IsDoctorMandatory: boolean = false;//sud:6Feb'18--for ward billing, used only in client side.
  public IsZeroPriceAllowed: boolean = false;//pratik: 2Feb'21--for Zero Price Allowed billitems, used only in client side.
  public IsPriceValid: boolean = true;//pratik: 12 May,2021 --chech validation on price, highlight price if  IsZeroPriceAllowed false and price is zero
  //Price Category includes: Normal, Foreigner, EHS, SAARCCitizen, GovtInsurance, etc.. 
  public PriceCategory: string = "Normal";//sud: 25Feb'19-- default value is Normal

  //checked only if we need to do a post to INS_TXN_PatientInsurancePackages table
  public IsInsurancePackage: boolean = false;
  public PatientInsurancePackageId: number = null;
  public FullName: string = "";           //ag7_mig_fix: used in calculate-details.component.html page.
  public StartDate: string = null;
  public EndDate: string = null;
  public IsExistBedFeatureId: boolean = false;
  public ModifiedBy: number = null;
  public ModifiedOn: string = null;
  public IsLastBed: boolean = false;
  public ProvisionalReceiptNo: number = null; //Yubraj 31st May '19
  public ProvisionalFiscalYearId: number = null; //Yubraj 31st May '19
  public ProvisionalFiscalYear: number = null; //Yubraj 31st May '19
  public IsInsurance: boolean = false;//Yubraj 3rd July '19
  public DiscountSchemeId: number = null; //Yubraj 30th July '19

  public DiscountApplicable: boolean = true;//sud:4Sept'19--only in client side

  //in cas of discount edit, if we have to cancel the edit then we should take oldDiscountPerecent.
  public OldDiscountPercent: number = 0;//sud:6Sept'19--Only in client side.

  public DisableAssignedDrField: boolean = false;//sud:1Oct'19--Only in client side.

  public AllowMultipleQty: boolean = true;
  public BillingTransaction: BillingTransaction = null;//sud: 10Nov'19--to be used in billreturn+copy earlier items.
  public AssignedDoctorList: Array<any> = [];

  //pratik: 17Apr'20-- only for client side, to compare and show warning in billingtransaction page for double entry.
  public IsDoubleEntry_Now: boolean = false;
  public IsDoubleEntry_Past: boolean = false;

  public CreatedByObj = { EmployeeId: null, FullName: null, DepartmentName: null, };//Client side only
  public ModifiedByObj = { EmployeeId: null, FullName: null, DepartmentName: null, };//Client side only
  public DocObj: any = { EmployeeId: null, FullName: '' };//only for client side.

  public OrderStatus: string = null;//pratik: 7 Aug 2020
  public AllowCancellation: boolean = false;//Anish: 14 Aug 2020

  public LabTypeName: string = 'op-lab';//default value for lab-typename.
  public GovtInsurancePrice: number = 0;
  public InsBillItemPriceEditable : boolean = false; //aniket: 25 Mar 21- only for client side

  public ShowProviderName : boolean = false; //To show or hide provider name in the invoice.
  public IsValidIPItemLevelDisocunt : boolean = true; //To validate the item level discount percent in IP billing..
  constructor() {
    var _formBuilder = new FormBuilder();
    this.BillingTransactionItemValidator = _formBuilder.group({
      'ItemName': ['', Validators.compose([Validators.required])],
      'ServiceDepartmentId': ['', Validators.compose([Validators.required])],
      'RequestedBy': ['', Validators.compose([Validators.required])],
      //'RequestedBy': ['', Validators.compose([])],  //for biling order.
      'ProviderId': ['', Validators.compose([])],//there will be no validation for providerid at the begining. it is conditional validation.
      'Price': ['', Validators.compose([this.positiveNumberValdiator])],
      'Quantity': ['', Validators.compose([])],//its validator are conditional, and gets composed at runtime.
      'DiscountPercent': ['', Validators.compose([this.discountPercentValidator])],
    });
  }

  public IsDirty(fieldName): boolean {
    if (fieldName == undefined)
      return this.BillingTransactionItemValidator.dirty;
    else
      return this.BillingTransactionItemValidator.controls[fieldName].dirty;
  }

  //public IsValid(): boolean {
  //  if (this.BillingTransactionItemValidator.valid) {
  //    return true;
  //  }
  //  else {
  //    return false;
  //  }
  //}

  public IsValid(fieldName, validator): boolean {
    if (fieldName == undefined)
      return this.BillingTransactionItemValidator.valid;
    else
      return !(this.BillingTransactionItemValidator.hasError(validator, fieldName));
  }

  public IsValidCheck(fieldName, validator): boolean {
    if (fieldName == undefined)
      return this.BillingTransactionItemValidator.valid;
    else
      return !(this.BillingTransactionItemValidator.hasError(validator, fieldName));
  }

  //dynamically sets ON and OFF the validation on ProviderId controlname.
  public UpdateValidator(onOff: string, formControlName: string, validatorType: string) {
    let validator = null;
    if (validatorType == 'required' && onOff == "on") {
      validator = Validators.compose([Validators.required]);
    }
    else if(validatorType=="invalidNumber" && onOff=="on"){
      validator =  Validators.compose([this.positiveNumberValdiator]);
    }
    else {
      validator = Validators.compose([]);
    }

    this.BillingTransactionItemValidator.controls[formControlName].validator = validator;
    this.BillingTransactionItemValidator.controls[formControlName].updateValueAndValidity();

  }


  ////dynamically sets ON and OFF the validation on ProviderId controlname.
  //public ComposeValidators(formControlName: string, validatorTypes: Array<string>) {

  //  let validator = null;

  //  let validatorArr: Array<ValidatorFn> = [];

  //  if (validatorTypes && validatorTypes.length > 0) {

  //    validatorTypes.forEach(curValTypeName => {
  //      if (curValTypeName == "required") {
  //        validatorArr.push(Validators.required);
  //      }
  //      if (curValTypeName == "multipleQty") {
  //        validatorArr.push(this.multipleQtyValidator);
  //      }
  //      if (curValTypeName == "positiveNumber") {
  //        validatorArr.push(this.positiveNumberValdiator);
  //      }
  //    });

  //  }

  //  validator = Validators.compose(validatorArr);

  //  this.BillingTransactionItemValidator.controls[formControlName].validator = validator;
  //  this.BillingTransactionItemValidator.controls[formControlName].updateValueAndValidity();

  //}

  ///2017-07-19--For Quick Bug fix... Make proper implementation of it later..
  public static GetClone(ipBilTxnItm: BillingTransactionItem): BillingTransactionItem {


    //let srvDptId = ipBilTxnItm ? ipBilTxnItm.ServiceDepartment ? ipBilTxnItm.ServiceDepartment


    let retObject = Object.assign({}, ipBilTxnItm);
    //set not-required fields to null.--needs revision on which all fields to remove.
    retObject.Patient = null;
    retObject.ServiceDepartment = null;
    return retObject;
  }

  positiveNumberValdiator(control: FormControl): { [key: string]: boolean } {
    if (control) {
      if (control.value <= 0)
        return { 'invalidNumber': true };
    }

  }

  multipleQtyValidator(control: FormControl): { [key: string]: boolean } {
    if (control) {
      if (control.value > 1)
        return { 'invalidQty': true };
    }

  }


  discountPercentValidator(control: FormControl): { [key: string]: boolean } {
    if (control.value) {
      if (control.value < 0 || control.value > 100)
        return { 'invalidPercent': true };
    }

  }

  //to dynamically enable/disable any form-control. 
  //here [disabled] attribute was not working from cshtml, so written a separate logic to do it.
  public EnableControl(formControlName: string, enabled: boolean) {

    let currCtrol = this.BillingTransactionItemValidator.controls[formControlName];
    if (currCtrol) {
      if (enabled) {
        currCtrol.enable();
      }
      else {
        currCtrol.disable();
      }
    }
  }

  //remove individual validators of current billingtransactionitem. 
  //USAGE:
  ////let currBilTxnItm: BillingTransactionItem = new BillingTransactionItem();
  ////currBilTxnItm.RemoveValidators(["ProviderId","DiscountPercent"]);
  public RemoveValidators(formControls: Array<string>) {
    let validator = null;
    if (formControls != null && formControls.length > 0) {
      formControls.forEach(ctrlName => {
        this.BillingTransactionItemValidator.controls[ctrlName].validator = null;//Validators.compose([])
      });
    }
  }

}

File: Code/Components/DanpheEMR.ServerModel/BillingModels/BillingTransactionItemModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DanpheEMR.ServerModel
{
    public class BillingTransactionItemModel  // Its the model of the BIL_TXN_BillTransactionItems
    {
        [Key]
        public int BillingTransactionItemId { get; set; }
        [ForeignKey("BillingTransaction")]
        public int? BillingTransactionId { get; set; }
        public int PatientId { get; set; }
        public int? ProviderId { get; set; }
        public string ProviderName { get; set; }
        public int ServiceDepartmentId { get; set; }
        public string ServiceDepartmentName { get; set; }
        public string ProcedureCode { get; set; }
        public int ItemId { get; set; }
        public string ItemName { get; set; }
        public double Price { get; set; }
        public double? Quantity { get; set; }
        public double? SubTotal { get; set; }
        public double? DiscountPercent { get; set; }
        public double? DiscountPercentAgg { get; set; }
        public double? DiscountAmount { get; set; }
        public double? Tax { get; set; }
        public double? TotalAmount { get; set; }
        public string BillStatus { get; set; }
        //this requisitionid comes from other modules for Requisition. 
        public Int64? RequisitionId { get; set; }
        public DateTime? RequisitionDate { get; set; }
        public DateTime? CounterDay { get; set; }
        public int? CounterId { get; set; }
        public DateTime? PaidDate { get; set; }
        public bool? ReturnStatus { get; set; }
        public double? ReturnQuantity { get; set; }
        //this will be employeeid of the current user--sudarshan:7may'17
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public string Remarks { get; set; }
        public string CancelRemarks { get; set; }
        public double? TaxPercent { get; set; }
        public DateTime? CancelledOn { get; set; }
        //added sudarshan: 27May'17--to track who cancelled the credit item.
        public int? CancelledBy { get; set; }
        public int? RequestedBy { get; set; }
        public int? PatientVisitId { get; set; }
        public int? BillingPackageId { get; set; }

        public double? TaxableAmount { get; set; }

        public double? NonTaxableAmount { get; set; }//added: sud: 29May'18
        public int? PaymentReceivedBy { get; set; }//added: sud: 29May'18
        public int? PaidCounterId { get; set; }//added: sud: 29May'18       

        public string BillingType { get; set; }//added: sud: 19jun'18
        public int? RequestingDeptId { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }

        [NotMapped]
        public bool? IsTaxApplicable { get; set; } //added: ashim: 30May'18 to maintain TaxApplicable Property in client side during normal billing transaction

        //checked only if we need to do a post to INS_TXN_PatientInsurancePackages table
        [NotMapped]
        public bool IsInsurancePackage { get; set; }
        public int? PatientInsurancePackageId { get; set; }

        public virtual PatientModel Patient { get; set; }
        public virtual BillingTransactionModel BillingTransaction { get; set; }
        public virtual ServiceDepartmentModel ServiceDepartment { get; set; }


        public string VisitType { get; set; }//sud: 28Sept'18--needed for Lab, etc..

        [NotMapped]
        public string RequestingUserName { get; set; }//sud:24Sept'18
        [NotMapped]
        public string RequestingUserDept { get; set; }//sud:24Sept'18

        public static BillingTransactionItemModel GetClone(BillingTransactionItemModel ipTxnItem)
        {
            BillingTransactionItemModel retTxnItem = (BillingTransactionItemModel)ipTxnItem.MemberwiseClone();
            return retTxnItem;
        }


        //sud: 9Sept'18-- added for BillCancelltion feature.
        [NotMapped]
        public int CancelQty { get; set; }
        [NotMapped]
        public int ReturnQty { get; set; }
        [NotMapped]
        public int BalanceQty { get; set; }
        [NotMapped]
        public string ItemIntegrationName { get; set; }
        [NotMapped]
        public string SrvDeptIntegrationName { get; set; }

        [NotMapped]
        public string TransactionType { get; set; }

        //Price Category includes: Normal, Foreigner, EHS, SAARCCitizen, GovtInsurance, etc.. 
        public string PriceCategory { get; set; }//sud: 25Feb'19

        [NotMapped]
        public DateTime? StartedOn { get; set; }
        [NotMapped]
        public DateTime? EndedOn { get; set; }
        [NotMapped]
        public bool IsPatTransfered { get; set; }

        public int? ProvisionalReceiptNo { get; set; }//Yubraj 31th may '19
        public int? ProvisionalFiscalYearId { get; set; } //Yubraj 31th may '19

        [NotMapped]
        public string ProvFiscalYear { get; set; }

        [NotMapped]
        public bool IsLastBed { get; set; }

        public bool? IsInsurance { get; set; }

        public int? DiscountSchemeId { get; set; } //Yubraj 30th July '19

    }
}
